### PR TITLE
Added PDIRK and PRK solvers

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

solves a part of #2177

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Where are the tests for these ones?

**Comment by ParamThakkar123** [comment]
> Where are the tests for these ones?

I didn't see a separate file of tests for these ones but the tests are there in ode_convergence_tests.jl. The convergence test function for PDIRK44 is present inside Implicit Solver Convergence Tests and tests for KuttaPRK2p5 are present in Explicit Solver Convergence Tests

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas any changes to be made in the tests for these solvers ??

**Comment by ChrisRackauckas** [comment]
There is a test failure https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/10133887447/job/28019521857?pr=2303#step:6:953

File: lib/OrdinaryDiffEqPRK/src/OrdinaryDiffEqPRK.jl [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module OrdinaryDiffEqPRK
+
+import OrdinaryDiffEq: OrdinaryDiffEqAlgorithm, alg_order, OrdinaryDiffEqMutableCache,
+                       OrdinaryDiffEqConstantCache, constvalue, @unpack, @cache, alg_cache,
+                       unwrap_alg, perform_step!, @threaded
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqPRK/src/OrdinaryDiffEqPRK.jl:5]:
```suggestion
                       unwrap_alg, perform_step!, @threaded, initialize!
```

File: lib/OrdinaryDiffEqPDIRK/src/OrdinaryDiffEqPDIRK.jl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module OrdinaryDiffEqPDIRK
+
+import OrdinaryDiffEq: isfsal, alg_order, _unwrap_val, DEFAULT_PRECS, NLNewton,
+                       OrdinaryDiffEqNewtonAlgorithm, dolinsolve, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqMutableCache, constvalue, alg_cache, build_nlsolver,
+                       uses_uprev, nlsolve!, nlsolvefail, @unpack, unwrap_alg, @cache,
+                       markfirststage!, @threaded
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqPDIRK/src/OrdinaryDiffEqPDIRK.jl:7]:
```suggestion
                       markfirststage!, @threaded, initialize!, perform_step!
```

**Comment by ChrisRackauckas** [comment]
@oscardssmith https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/10157523456/job/28087785439?pr=2303#step:6:658 if we don't find a fix soon we'll need to revert the StaticArray change

**Comment by oscardssmith** [comment]
investigating now.

**Comment by oscardssmith** [comment]
From initial inspection, this doesn't seem related to my changes since it is happening in `addsteps!` which is pretty far away in the code from anything with the jacobians...

**Comment by ChrisRackauckas** [comment]
It's the ArrayInterface.jl changes.

**Comment by oscardssmith** [comment]
Ah, the problem is that `ArrayInterface.ismutable` is a generated function that illegally depends on world age I think.

```
>121  @generated function ArrayInterface.ismutable(::Type{<:ArrayPartition{T, S}}) where {T, S
 122  }
 123      res = all(ArrayInterface.ismutable, S.parameters)
 124      return :($res)
 125  end
```

**Comment by oscardssmith** [comment]
problem found. Partial revert incoming.

**Pull request merged**